plugins/targetlocker/dblocker/dblocker.go

```diff
@@ -0,0 +1,310 @@
+// (lock owner is ignored)
+func (d *DBLocker) cleanExpired(tx *sql.Tx, targets []string, now time.Time) error {
+	q := "DELETE FROM locks WHERE expires_at < ? AND target_id IN " + listQueryString(len(targets)) + ";"
+	// convert targets to a list of interface{}
+	queryList := make([]interface{}, 0, len(targets)+1)
+	queryList = append(queryList, now)
+	for _, targetId := range targets {
+		queryList = append(queryList, targetId)
+	}
+	_, err := tx.Exec(q, queryList...)
+	if err != nil {
+		return fmt.Errorf("unable to clean existing locks: %w", err)
+	}
+	return nil
+}
+
+// queryLocks returns a map of ID -> dblock for a given list of targets
+func (d *DBLocker) queryLocks(tx *sql.Tx, targets []string) (map[string]dblock, error) {
+	q := "SELECT target_id, job_id, expires_at FROM locks WHERE target_id IN " + listQueryString(len(targets))
+	// convert targets to a list of interface{}
+	queryList := make([]interface{}, 0, len(targets))
+	for _, targetId := range targets {
+		queryList = append(queryList, targetId)
+	}
+
+	rows, err := tx.Query(q, queryList...)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read existing locks: %w", err)
+	}
+	defer rows.Close()
+
+	row := dblock{}
+	locks := make(map[string]dblock)
+	for rows.Next() {
+		if err := rows.Scan(&row.targetId, &row.jobID, &row.expiresAt); err != nil {
+			return nil, fmt.Errorf("unexpected read from database: %w", err)
+		}
+		locks[row.targetId] = row
+	}
+	return locks, nil
+}
+
+// handleLock does the real locking, it assumes the jobID is valid
+func (d *DBLocker) handleLock(jobID int64, targets []string, timeout time.Duration) error {
+	// everything operates on this frozen time
+	now := time.Now()
+	expires := now.Add(timeout)
+	// locking is all or nothing in one transaction
+	// phantom reads might mess with the expired lock cleaning assumptions below,
+	// so request serializable isolation
+	tx, err := d.db.BeginTx(context.Background(), &sql.TxOptions{Isolation: sql.LevelSerializable})
+	if err != nil {
+		return fmt.Errorf("unable to start database transaction: %w", err)
+	}
+	defer func() {
+		// this always fails if tx.Commit() was called before, ignore error
+		_ = tx.Rollback()
+	}()
+
+	// clean expired locks first, this simplifies the logic later
+	if err := d.cleanExpired(tx, targets, now); err != nil {
+		return err
+	}
+
+	locks, err := d.queryLocks(tx, targets)
+	if err != nil {
+		return err
+	}
+	// go through existing locks, they are either held by something else (abort)
+	// help by us (update time), or not held (insert)
+	inserts := make([]string, 0)
+	updates := make([]string, 0)
+	conflicts := make([]dblock, 0)
+	for _, t := range targets {
+		lock, ok := locks[t]
+		switch {
+		case !ok:
+			inserts = append(inserts, t)
+		case lock.jobID == jobID:
+			updates = append(updates, t)
+		default:
+			conflicts = append(conflicts, lock)
+		}
+	}
+	if len(conflicts) > 0 {
+		return fmt.Errorf("unable to lock, targets have conflicting locks: %v", conflicts)
```

Comment:
we probably also want the job ID and the target ID in this error string, for sake of debugging when necessary

Reply:
They are! This is why there is a pretty printer on the dblock struct :)

Reply:
I mean not just the info on what is currently holding the lock, but also on what tried to get the lock but failed

Reply:
done, looks like this now:
```
unable to lock targets [001 002] for owner 124, have conflicting locks: [target: 001 job: 123 expires: 2020-08-25 15:01:48 +0000 UTC target: 002 job: 123 expires: 2020-08-25 15:01:48 +0000 UTC]
```